### PR TITLE
Fixes #2 by introcution of embedded Hsqldb tests

### LINKED ISSUE #2: Use embedded DB for test cases
Originally reported on Google Code with ID 2

```
Currently the JWPL test cases require access to our DB host. This precludes
external users from running the tests (when JWPL is open sourced) and it also
prevents tests when the DB host is offline (as currently...).

Instead an embedded DB like HSQLDB or Derby should be used for test cases.


```

Reported by `torsten.zesch` on 2010-09-21 16:07:20

### PR DESCRIPTION
**Pull request by mawiesne**
Please review changes and consider merging this PR against master branch.

Carefully check the current branch with a MySQL version of the database by adjusting configuration properties in `BaseJWPLTest`.

Feedback welcome.

### CODE REVIEW

**Comment by reckart** [comment]
Jenkins, can you test this please?

**Comment by reckart** [comment]
Note: I have to trigger this build again later today as the CI is restarted soon and currently in shutdown mode.

**Comment by mawiesne** [comment]
@reckart It is good that your CI is down ;-). I forgot to check other modules than the API one. Thus, I will have to provide ported hsqldb setup to non-api modules and add 1 or 2 more commits. pls hold the line with Jenkins

**Comment by reckart** [comment]
@mawiesne Could you please push some change to this branch so we can test if the whitelisting worked?

**Comment by mawiesne** [comment]
Jenkins, can you test this please?

**Comment by reckart** [comment]
@mawiesne looks like Jenkins heard you :)

**Comment by mawiesne** [comment]
@reckart Seems fine. I will continue work on this PR later this evening and/or tomorrow.

**Comment by reckart** [comment]
@mawiesne At some point, please re-write the commit comments in this PR to the following format:

```
#<ISSUE NUMBER> - <ISSUE TITLE>
<EMPTY LINE>
- <CHANGE 1>
- <CHANGE 2>
- <...>
```

Note that `#` normally introduces comments in git. You may have to reconfigure git before attempting an interactive rebase and switch it to another comment character: https://stackoverflow.com/questions/22936252/escape-comment-character-in-git-commit-message

File: de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/DatabaseConfiguration.java [hunk 1]
```diff
@@ -39,6 +41,20 @@ public DatabaseConfiguration(String host, String database, String user, String p
...
+        // static mysql usecase - default by revision 49
+        this.setDatabaseDriver("com.mysql.jdbc.Driver");
+        this.setJdbcURL("jdbc:mysql://" + host + "/" + database);
+    }
+
+    /**
+     * Allows explicit DMBS type specific configuration for hsqldb from junit tests context
+     */
+    public DatabaseConfiguration(String databaseDriver, String jdbcURL, String host, String database, String user, String password, Language language) {
+    	this(host, database, user, password, language);
```
Comment (tgalery) [inline de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/DatabaseConfiguration.java]: funny indent

File: de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/DatabaseConfiguration.java [hunk 2]
```diff
@@ -39,6 +41,20 @@ public DatabaseConfiguration(String host, String database, String user, String p
         this.user = user;
         this.password = password;
         this.language = language;
+
+
+        // static mysql usecase - default by revision 49
+        this.setDatabaseDriver("com.mysql.jdbc.Driver");
+        this.setJdbcURL("jdbc:mysql://" + host + "/" + database);
```
Comment (tgalery) [inline de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/DatabaseConfiguration.java]: would the mysql prefix here work with the embedded db ?

File: de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/hibernate/WikiHibernateUtil.java [hunk 3]
```diff
@@ -56,48 +56,83 @@ else if (config.getDatabase() == null) {
...
 
+        // Database connection settings
+        p.setProperty("hibernate.connection.driver_class", databaseDriverClass);
+        p.setProperty("hibernate.connection.url", jdbcURL);
+        /*
+         *  Needed to ensure working hsqldb queries - don't remove it...!
+         */
+        p.setProperty("hibernate.connection.useUnicode","true");
         p.setProperty("hibernate.connection.characterEncoding", "UTF-8");
+//        p.setProperty("hibernate.connection.hibernate.connection.characterEncoding", "UTF-8");
```
Comment (tgalery) [inline de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/hibernate/WikiHibernateUtil.java]: remove commented out lines

File: de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java [hunk 4]
```diff
@@ -28,31 +28,24 @@
...
-		try {
-			wiki = new Wikipedia(db);
-		} catch (Exception e) {
-			assumeNoException(e);
-			//fail("Wikipedia could not be initialized.");
-		}
+        DatabaseConfiguration db = obtainHSDLDBConfiguration();
+        try {
+            wiki = new Wikipedia(db);
+        } catch (Exception e) {
```
Comment (tgalery) [inline de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java]: I wonder whether we should remove these try - catches as they might create the illusion that all tests are passing (specially for new contributors)

**Comment by tgalery** [comment]
Just double checking: there some scripts that are used to pre populate the embedded database so tests actually run. Does the library for the embedded database runs the scripts or do we need to change something in the configuration of jenkins so the test infrastructure can run the scripts before the tests themselves

**Comment by mawiesne** [comment]
@tgalery HSQL bootstraps the DB via the scripts. There are not further changes needed for the build environment. Jenkins setup in Darmstadt seems to work fine with it. Still, the current PR is work in progress. Thanks for your comments. Will work on it during this week.

File: de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/PageIteratorTest.java [hunk 5]
```diff
@@ -18,38 +18,29 @@
...
 	 */
 	@BeforeClass
 	public static void setupWikipedia() {
-		DatabaseConfiguration db = new DatabaseConfiguration();
-		db.setDatabase("wikiapi_test");
-		db.setHost("bender.ukp.informatik.tu-darmstadt.de");
-		db.setUser("student");
-		db.setPassword("student");
-		db.setLanguage(Language._test);
+		DatabaseConfiguration db = obtainHSDLDBConfiguration();
```
[the remark this answers is not in the document]
Comment (mawiesne) [reply de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/PageIteratorTest.java]: The code in `obtainHSDLDBConfiguration()` in `BaseJWPLTest` avoids code duplication and is re-used in several test clasess. That's why I've extracted it. I will introduce different Maven profiles to exchange hard coded parameters in `obtainHSDLDBConfiguration()` dynamically from a  devops' perspective.

File: de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java (hunk 4 above)
Comment (mawiesne) [reply de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java]: I don't see why these are creating an illusion here. During test setup phase, exceptions might occur and that's why there is a try/catch here. Maybe, I don't get your concern here correctly.

File: de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/DatabaseConfiguration.java (hunk 2 above)
Comment (mawiesne) [reply de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/DatabaseConfiguration.java]: You're correct. This is a "special" constructor atm, as this is used by production code fragments whilst the other (new) constructor is only here for making HSQL embedded setup work. I'll change this with the next commits to come.

File: de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/DatabaseConfiguration.java (hunk 1 above)
Comment (mawiesne) [reply de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/DatabaseConfiguration.java]: Correct, will fix it.

File: de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java (hunk 4 above)
Comment (reckart) [reply de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java]: @mawiesne Unit tests should normally not contain try/catch statements. If an exception is thrown, JUnit should pick it up and mark the test as in error. The only reason to catch an exception would normally be to check its type/error message.

Comment (mawiesne) [reply de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java]: @reckart Btw, this was the original code of the master branch, merely adapted. Shall I change it throughout all the other test cases in the api module?

Comment (rzo1) [reply de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java]: And a reason for #161 ...

File: de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/hibernate/WikiHibernateUtil.java (hunk 3 above)
Comment (mawiesne) [reply de.tudarmstadt.ukp.wikipedia.api/src/main/java/de/tudarmstadt/ukp/wikipedia/api/hibernate/WikiHibernateUtil.java]: +1

File: de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java (hunk 4 above)
Comment (reckart) [reply de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java]: @mawiesne up to you, but if you do, I'd recommend doing it in a separate PR.

Comment (mawiesne) [reply de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java]: @reckart you're right, code style is a topic to watch for. Once we have working (embedded) tests and this PR is merged, we can discuss/address changes related to this topic within #161. 
I see this PR as a first step in a row of pre-release actions. As already mentioned by @rzo1, testability should be improved, with it: code, code documentation, reduction of open TODOs/FIXMEs...

**Comment by mawiesne** [comment]
@daxenberger @reckart Basically, I'm done with fixing #2. 

With the lastest commit, no existing JUnit test (class) depends on `bender.ukp.informatik.tu-darmstadt.de` at runtime. The only remaining code fragments that use this URL are found in "main"-methods which are no true tests... 

Please review the changes and merge the PR if those are fine with you. Please check/consider the chain of steps mentioned in comments of #158.  If you signal approval, either you or me can hit the Merge PR button. 

@rzo1 / @tgalery: Your feedback is also welcome.

File: de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java (hunk 4 above)
Comment (reckart) [reply de.tudarmstadt.ukp.wikipedia.api/src/test/java/de/tudarmstadt/ukp/wikipedia/api/CategoryDescendantsIteratorTest.java]: @mawiesne wrt style cf https://github.com/dkpro/dkpro-jwpl/issues/164

**Comment by mawiesne** [comment]
@reckart I tried to do so, but IntelliJ ran into trouble with the older commit (pre 2018) and failed...

**Comment by reckart** [comment]
@mawiesne I usually do this on the command-line.

**Comment by mawiesne** [comment]
@reckart For future commits, I'll be fine with this policy.

In this context: Could you put this important information for committers somewhere in a central contribution guide / doc / FAQ? This c/would avoid that we run into the same commit-message style topic with other devs again.

**Comment by reckart** [comment]
@mawiesne I have added instructions on doing pull requests and formatting commits into the DKPro contribution guidelines: https://dkpro.github.io/contributing/

**Comment by reckart** [comment]
@mawiesne I can try rewriting the commit comments - however, it will require that I do a force-push into this branch. That means if anybody else who has checked out that branch (i.e. you) needs to locally drop the branch and check it out again before they can make any updates to the branch. Is that ok for you? Can I rewrite the messages?

**Comment by mawiesne** [comment]
@reckart Sure, go for it.

**Comment by reckart** [comment]
@mawiesne unfortunately, commandline git is also having trouble with the interactive rebase but it doesn't say in detail why. So let's live with the commit messages as they are then.

**Comment by mawiesne** [comment]
@reckart Told you. Thx for trying. Let's proceed with the merge then.

**Pull request merged**